### PR TITLE
Upgrades `clap` from 2.x to 4.x

### PR DESCRIPTION
**Pull request by zslayton**
The `ion-cli` now uses the latest major version of `clap`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,7 +13,7 @@ keywords = ["format", "parse", "encode"]
 
 [dependencies]
 anyhow = "1.0"
-clap = "~2.27.0"
+clap = { version = "4.0.17", features = ["cargo"] }
```
Comment (zslayton) [inline Cargo.toml:16]: 🗺️ The `cargo` feature allows `clap` to pull information about the crate from its `Cargo.toml` file at compile time. This makes it possible to include the list of crate authors and the crate version in the `--help` message.

File: src/bin/ion/commands/beta/count.rs [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
-use clap::{App, Arg, ArgMatches};
+use clap::{Arg, ArgAction, ArgMatches, Command};
 use ion_rs::*;
 use std::fs::File;
 use std::io::{stdin, BufReader, StdinLock};
 
-pub fn app() -> CommandConfig {
-    App::new("count")
+pub fn app() -> Command {
+    Command::new("count")
```
Comment (zslayton) [inline src/bin/ion/commands/beta/count.rs:8]: 🗺️ `clap` renamed its `App` construct to `Command`.

File: src/bin/ion/commands/beta/count.rs [hunk 3]
```diff
@@ -1,25 +1,25 @@
...
             // All argv entries after the program name (argv[0])
             // and any `clap`-managed options are considered input files.
-            Arg::with_name("input")
+            Arg::new("input")
                 .index(1)
-                .multiple(true)
-                .help("Input file [default: STDIN]"),
+                .help("Input file [default: STDIN]")
+                .action(ArgAction::Append)
+                .trailing_var_arg(true),
```
Comment (zslayton) [inline src/bin/ion/commands/beta/count.rs:17]: 🗺️ The `multiple(true)` setting has been removed. You can get the same effect by changing the `ArgAction` associated with the arg name from `Set` to `Append`.

Setting `trailing_var_arg` to `true` indicates that all trailing arguments not associated with a flag will be part of `input`.

File: src/bin/ion/commands/beta/count.rs [hunk 4]
```diff
@@ -1,25 +1,25 @@
...
         )
 }
 
-pub fn run(_command_name: &str, matches: &ArgMatches<'static>) -> Result<()> {
-    if let Some(input_file_iter) = matches.values_of("input") {
+pub fn run(_command_name: &str, matches: &ArgMatches) -> Result<()> {
+    if let Some(input_file_iter) = matches.get_many::<String>("input") {
```
Comment (zslayton) [inline src/bin/ion/commands/beta/count.rs:22]: 🗺️ `clap` added more robust facilities for parsing a given CLI argument's text into a specific type. In our case, we actually want a `String` most of the time so it's not buying us much.

File: src/bin/ion/commands/beta/count.rs [hunk 5]
```diff
@@ -1,25 +1,25 @@
...
                 .index(1)
-                .multiple(true)
-                .help("Input file [default: STDIN]"),
+                .help("Input file [default: STDIN]")
+                .action(ArgAction::Append)
+                .trailing_var_arg(true),
         )
 }
 
-pub fn run(_command_name: &str, matches: &ArgMatches<'static>) -> Result<()> {
```
Comment (zslayton) [inline src/bin/ion/commands/beta/count.rs:21]: 🗺️ `clap` was able to remove an extraneous lifetime from `ArgMatches`.

File: src/bin/ion/commands/beta/inspect.rs [hunk 6]
```diff
@@ -19,31 +19,31 @@ const ABOUT: &str =
...
+pub fn app() -> Command {
+    Command::new("inspect")
         .about(ABOUT)
         .arg(
-            Arg::with_name("output")
+            Arg::new("output")
                 .long("output")
-                .short("o")
-                .takes_value(true)
+                .short('o')
```
Comment (zslayton) [inline src/bin/ion/commands/beta/inspect.rs:28]: 🗺️ The `short` name method takes a `char` instead of a `&str`. I didn't investigate the motivation.

**Review by zslayton — COMMENTED** [review]
🗺️ Code tour

**Review by desaikd — APPROVED** [review]
LGTM!

**Pull request merged**